tools/pubsys/src/aws/validate_ami/mod.rs

```diff
@@ -0,0 +1,799 @@
+                    region.clone(),
+                    images.iter().map(|i| i.id.clone()).collect::<Vec<String>>(),
+                )
+            })
+            .collect::<HashMap<Region, Vec<String>>>(),
+        &expected_image_public,
+    )
+    .await;
+
+    // Validate the retrieved EC2 images per region
+    info!("Validating EC2 images");
+    let results: HashMap<Region, ami::Result<HashSet<AmiValidationResult>>> = images
+        .into_iter()
+        .map(|(region, region_result)| {
+            (
+                region.clone(),
+                region_result.map(|result| {
+                    validate_images_in_region(
+                        expected_images
+                            .get(region)
+                            .map(|e| e.to_owned())
+                            .unwrap_or(vec![]),
+                        &result,
+                        region,
+                    )
+                }),
+            )
+        })
+        .collect::<HashMap<Region, ami::Result<HashSet<AmiValidationResult>>>>();
+
+    let validation_results = AmiValidationResults::new(results);
+
+    // If a path was given to write the results to, write the results
+    if let Some(write_results_path) = &validate_ami_args.write_results_path {
+        // Filter the results by given status, and if no statuses were given, get all results
+        info!("Writing results to file");
+        let filtered_results = validation_results.get_results_for_status(
+            validate_ami_args
+                .write_results_filter
+                .as_ref()
+                .unwrap_or(&vec![
+                    AmiValidationResultStatus::Correct,
+                    AmiValidationResultStatus::Incorrect,
+                    AmiValidationResultStatus::Missing,
+                ]),
+        );
+
+        // Write the results as JSON
+        serde_json::to_writer_pretty(
+            &File::create(write_results_path).context(error::WriteValidationResultsSnafu {
+                path: write_results_path,
+            })?,
+            &filtered_results,
+        )
+        .context(error::SerializeValidationResultsSnafu)?;
+    }
+
+    Ok(validation_results)
+}
+
+/// Validates EC2 images in a single region, based on a Vec (ImageDef) of expected images
+/// and a HashMap (AmiId, ImageDef) of actual retrieved images. Returns a HashSet of
+/// AmiValidationResult objects.
+pub(crate) fn validate_images_in_region(
+    expected_images: Vec<ImageDef>,
+    actual_images: &HashMap<AmiId, ImageDef>,
+    region: &Region,
+) -> HashSet<AmiValidationResult> {
+    let mut results = HashSet::new();
+
+    // Validate all expected images, creating an AmiValidationResult object
+    for mut image in expected_images {
+        // If the image is expected to be public, the specific launch permissions are irrelevant
+        // and were not retrieved for the actual image
+        if image.public {
+            image.launch_permissions = None;
+        }
```

Comment:
Why isn't `image.launch_permissions` set to None when the `ImageDef` is created? If that's not possible can `launch_permissions` be a function?